ipa-core/src/secret_sharing/replicated/semi_honest/additive_share.rs

```diff
@@ -86,138 +122,161 @@ where
     }
 }
 
-impl<'a, 'b, V: SharedValue> Add<&'b AdditiveShare<V>> for &'a AdditiveShare<V> {
-    type Output = AdditiveShare<V>;
+impl<'a, 'b, V: SharedValue + Vectorizable<N>, const N: usize> Add<&'b AdditiveShare<V, N>>
+    for &'a AdditiveShare<V, N>
+{
+    type Output = AdditiveShare<V, N>;
 
-    fn add(self, rhs: &'b AdditiveShare<V>) -> Self::Output {
-        AdditiveShare(self.0 + rhs.0, self.1 + rhs.1)
+    fn add(self, rhs: &'b AdditiveShare<V, N>) -> Self::Output {
+        AdditiveShare(
+            Add::add(self.0.clone(), &rhs.0),
```

Comment:
You may be thinking of the discussion about the clones in the `multiply` implementation, and the issue is the same -- this is dispatching via the `SharedValueArray` trait, which requires `Array: Sub<Array>` and `Array: Sub<&Array>`, but can't require `&Array: Sub<&Array>`, because it's not possible to express a bound on a reference to the type in a supertrait bound.

If we really wanted to avoid these clones, we could implement a custom trait that provides ref-by-ref arithmetic operations as an inherent method accepting two arguments (`trait MyAdd { fn add(a: &Self, b: &Self) -> Self; }`. I feel like the compiler is probably doing a pretty good job of optimizing these clones so that it may not be worth the trouble to do that, but I don't have solid data to back that up.

We could also drop the ref-by-ref impls here, which would make it more explicit in the caller when a clone is taking place.